src/linear_algebra/dual.lean

```diff
@@ -398,3 +407,156 @@ begin
+  simp_rw [linear_map.ext_iff, dual_restrict_apply],
+  exact ⟨λ h w hw, h ⟨w, hw⟩, λ h w, h w.1 w.2⟩
+end
+
+lemma dual_restrict_ker_eq_dual_annihilator
+  (W : submodule R M) : W.dual_restrict.ker = W.dual_annihilator :=
+rfl
+
+end submodule
+
+namespace subspace
+
+open submodule linear_map
+
+universes u v w
+
+-- We work in vector spaces because `exists_is_compl` only hold for vector spaces
+variables {K : Type u} {V : Type v} [field K] [add_comm_group V] [vector_space K V]
+
+/-- Given a subspace `W` of `V` and an element of its dual `φ`, `dual_lift W φ` is
+  the natural extenstion of `φ` to an element of the dual of `V`. -/
+noncomputable def dual_lift (W : subspace K V) :
+  module.dual K W →ₗ[K] module.dual K V :=
+let h := classical.indefinite_description _ W.exists_is_compl in
+  (linear_map.of_is_compl_prod h.2).comp (linear_map.inl _ _ _)
+
+variable {W : subspace K V}
+
+@[simp] lemma dual_lift_of_subtype {φ : module.dual K W} (w : W) :
+  W.dual_lift φ (w : V) = φ w :=
+by { erw of_is_compl_left_apply _ w, refl }
+
+lemma dual_lift_of_mem {φ : module.dual K W} {w : V} (hw : w ∈ W) :
+  W.dual_lift φ w = φ ⟨w, hw⟩ :=
+dual_lift_of_subtype ⟨w, hw⟩
+
+lemma dual_lift_left_inverse (W : subspace K V) :
+  W.dual_restrict.comp W.dual_lift = 1 :=
+by { ext φ x, simp }
+
+lemma dual_restrict_surjective :
+  function.surjective W.dual_restrict :=
+begin
+  intros φ, refine ⟨W.dual_lift φ, _⟩, ext,
+  erw [dual_restrict_apply, dual_lift, of_is_compl_left_apply], refl
+end
+
+lemma dual_lift_injective : function.injective W.dual_lift :=
+begin
+  rintro _ _ h,
+  ext, rw [← dual_lift_of_subtype, h, dual_lift_of_subtype],
+end
+
+/-- The quotient by the `dual_annihilator` of a subspace is isomorphic to the
+  dual of that subspace. -/
+noncomputable def quot_annihilator_equiv (W : subspace K V) :
+  W.dual_annihilator.quotient ≃ₗ[K] module.dual K W :=
+(quot_equiv_of_eq _ _ W.dual_restrict_ker_eq_dual_annihilator).symm.trans $
+  W.dual_restrict.quot_ker_equiv_of_surjective dual_restrict_surjective
+
+/-- The natural linear map from the dual of a subspace `W` to `W.dual_lift.range`. -/
+noncomputable def dual_to_subspace_dual (W : subspace K V) :
+  module.dual K W →ₗ[K] W.dual_lift.range :=
+W.dual_lift.range_restrict
+
+@[simp] lemma dual_to_subspace_dual_apply (φ : module.dual K W) :
+  W.dual_to_subspace_dual φ = ⟨W.dual_lift φ, mem_range.2 ⟨φ, rfl⟩⟩ := rfl
+
+lemma dual_to_subspace_ker_eq_bot :
+  W.dual_to_subspace_dual.ker = ⊥ :=
+linear_map.ker_eq_bot.2 $ λ φ ψ h, dual_lift_injective (subtype.mk_eq_mk.1 h)
+
+lemma dual_to_subspace_range_eq_top :
+  W.dual_to_subspace_dual.range = ⊤ :=
+W.dual_lift.range_range_restrict
+
+/-- The natural isomorphism forom the dual of a subspace `W` to `W.dual_lift.range`. -/
+noncomputable def dual_equiv_dual (W : subspace K V) :
+  module.dual K W ≃ₗ[K] W.dual_lift.range :=
+linear_equiv.of_bijective W.dual_to_subspace_dual
+  dual_to_subspace_ker_eq_bot dual_to_subspace_range_eq_top
```

Comment:
You can avoid needing `dual_to_subspace_range_eq_top` with  `linear_equiv.of_injective _ dual_to_subspace_ker_eq_bot`

Reply:
Right, I can also just replace it with `W.dual_lift.range_range_restrict` which is even shorter

Reply:
I don't follow - if you use `linear_equiv.of_injective` instead of `linear_equiv.of_bijective` then you don't need the statement `W.dual_to_subspace_dual.range = ⊤` at all

Reply:
Ah, I misunderstood. I see what you mean now

Reply:
It seems like I can write the isomorphism as `linear_equiv.of_injective _ $ ker_eq_bot.2 dual_lift_injective` so we do not even need the linear map `dual_to_subspace_dual` at all!